### ISSUE TITLE
Error when using any learning method except SGD.

### ISSUE THREAD

**Issue by sea2603** [glample/tagger#3]
Good work!
AttributeError when using learning methods except SGD. Does the error happen when try to update the word embedding?

**Comment by kaituoxu** [comment]
Have you solved this problem?

**Comment by sea2603** [comment]
no    (〜￣▽￣)〜

**Comment by glample** [comment]
Sorry.. it looks like the problem comes from the fact that I select a submatrix do to the word embeddings updates.

The idea was that when you load pretrained embeddings to test a model, you will only store in your lookup table the words that are relevant, ie. the words in the train / dev / test.
But if you want to train a model and use it for real applications, you will actually need to use the entire list of pretrained word embeddings. However, most of these words will not appear in the training set, and there is no point updating them during the training since their updates will always be 0.

This is the point of the argument "all_emb" in the training file. By default it's set to 0 (to test different model configurations), but if you want to train a model and use it, you need to set it to 1.

If you look at https://github.com/glample/tagger/blob/master/model.py#L376, this line basically indicates that we don't need to update the words that do not appear in the input. This is theoretically much faster, but in practice I didn't observe much difference with and without this.

Anyway, can you try to replace:

Line 379: `params.extend([word_input])` by: `params.extend(word_layer.params)`

And also, lines 373 - 378:

```
_updates = Optimization(clip=5.0).get_updates(lr_method_name, cost, params, **lr_method_parameters)
updates = [(word_layer.embeddings, T.set_subtensor(word_input, _updates[0][1]))]
updates += _updates[1:]
```

By just this:

```
updates = Optimization(clip=5.0).get_updates(lr_method_name, cost, params, **lr_method_parameters)
```

Please tell me if this works.

**Comment by sea2603** [comment]
Thanks.It works.

**Comment by kaituoxu** [comment]
Thanks. It works.

I have a suggestion, maybe you can make your code more efficient.
1. You can add parallel in the LSTM gates computing like this tutorial http://deeplearning.net/tutorial/lstm.html, in the **Implementation note** part.( Ctrl+F, 'Implementation note')
2. You can add minibatch training. You can read this lstm implement code and using mask matrix.  http://deeplearning.net/tutorial/lstm.html in the **Code** part.

I'm a newer to Theano, if you think my suggestion is naive, just ignore it.
: )

**Comment by glample** [comment]
Thanks for the suggestions!

Indeed, this is faster to make all operations in the LSTM at once. However, the trick they propose can't be fully applied with peephole connections (ie. it doesn't allow to take the previous value of the cell to compute the values of the next gates). And in practice I observed better results with these peephole connections.

But the difference with and without peephole is not a lot and it's probably worth using the implementation they have. However, although this trick is very efficient for code running on GPU, I never observed a significant difference when running it on CPU. And my code was running much faster on CPU (since the hidden layers are pretty small), so I didn't use it.

For the batching part, I am actually already using something similar to what they call "mask matrix" I guess, because I need to compute embeddings for each words by feeding their characters to a LSTM, and the words have different lengths. Now, to have several sentences by batch, you would need a 3D tensor just for the characters. This is a little painful in theano, but nothing impossible. Maybe something more difficuly would be to allow batches for the CRF layer. If you want to try, I can help you with that =)

**Comment by kaituoxu** [comment]
I have implemented the first suggestion I mentioned above, and there is a little improvement. : )

``` python
class LSTM(object):
    """
    Long short-term memory (LSTM). Can be used with or without batches.
    Without batches:
        Input: matrix of dimension (sequence_length, input_dim)
        Output: vector of dimension (output_dim)
    With batches:
        Input: tensor3 of dimension (batch_size, sequence_length, input_dim)
        Output: matrix of dimension (batch_size, output_dim)
    """
    def __init__(self, input_dim, hidden_dim, with_batch=True, name='LSTM'):
        """
        Initialize neural network.
        """
        self.input_dim = input_dim
        self.hidden_dim = hidden_dim
        self.with_batch = with_batch
        self.name = name

        # parallel
        self.w_x = shared((input_dim, 3 * hidden_dim), name + '__w_x')  #w_xi, w_xc, w_xo
        self.w_h = shared((hidden_dim, 3 * hidden_dim), name + '__w_h') #w_hi, w_hc, w_ho
        self.b = shared((3 * hidden_dim,), name + '__b')                #b_i, b_c, b_o,

        # Input gate weights
        ####self.w_xi = shared((input_dim, hidden_dim), name + '__w_xi')
        ####self.w_hi = shared((hidden_dim, hidden_dim), name + '__w_hi')
        self.w_ci = shared((hidden_dim, hidden_dim), name + '__w_ci')

        # Forget gate weights
        # self.w_xf = shared((input_dim, hidden_dim), name + '__w_xf')
        # self.w_hf = shared((hidden_dim, hidden_dim), name + '__w_hf')
        # self.w_cf = shared((hidden_dim, hidden_dim), name + '__w_cf')

        # Output gate weights
        ####self.w_xo = shared((input_dim, hidden_dim), name + '__w_xo')
        ####self.w_ho = shared((hidden_dim, hidden_dim), name + '__w_ho')
        self.w_co = shared((hidden_dim, hidden_dim), name + '__w_co')

        # Cell weights
        ####self.w_xc = shared((input_dim, hidden_dim), name + '__w_xc')
        ####self.w_hc = shared((hidden_dim, hidden_dim), name + '__w_hc')

        # Initialize the bias vectors, c_0 and h_0 to zero vectors
        ####self.b_i = shared((hidden_dim,), name + '__b_i')
        ##### self.b_f = shared((hidden_dim,), name + '__b_f')
        ####self.b_c = shared((hidden_dim,), name + '__b_c')
        ####self.b_o = shared((hidden_dim,), name + '__b_o')
        self.c_0 = shared((hidden_dim,), name + '__c_0')
        self.h_0 = shared((hidden_dim,), name + '__h_0')

        # Define parameters
        self.params = [self.w_x, self.w_ci,#####self.w_xi, self.w_hi, self.w_ci,
                       # self.w_xf, self.w_hf, self.w_cf,
                       self.w_h, self.w_co,#self.w_xo, self.w_ho, self.w_co,
                       ####self.w_xc, self.w_hc,
                       self.b,####self.b_i, self.b_c, self.b_o,  # self.b_f,
                       self.c_0, self.h_0]

    def _slice(self, _x, n, dim):
        if _x.ndim == 1:
            return _x[n * dim:(n + 1) * dim]

    def link(self, input):
        """
        Propagate the input through the network and return the last hidden
        vector. The whole sequence is also accessible via self.h, but
        where self.h of shape (sequence_length, batch_size, output_dim)
        """
        def recurrence(x_t, c_tm1, h_tm1):
            preact = T.dot(h_tm1, self.w_h)
            preact += x_t

        #import pdb; pdb.set_trace()

            i_t = T.nnet.sigmoid(self._slice(preact, 0, self.hidden_dim) + T.dot(c_tm1, self.w_ci))
            ####i_t = T.nnet.sigmoid(T.dot(x_t, self.w_xi) +
            ####                     T.dot(h_tm1, self.w_hi) +
            ####                     T.dot(c_tm1, self.w_ci) +
            ####                     self.b_i)
            # f_t = T.nnet.sigmoid(T.dot(x_t, self.w_xf) +
            #                      T.dot(h_tm1, self.w_hf) +
            #                      T.dot(c_tm1, self.w_cf) +
            #                      self.b_f)
            c_t = ((1 - i_t) * c_tm1 + i_t * T.tanh(self._slice(preact, 1, self.hidden_dim)))
            ####c_t = ((1 - i_t) * c_tm1 + i_t * T.tanh(T.dot(x_t, self.w_xc) +
            ####       T.dot(h_tm1, self.w_hc) + self.b_c))
            o_t = T.nnet.sigmoid(self._slice(preact, 2, self.hidden_dim) +
                                 T.dot(c_t, self.w_co))
            ####o_t = T.nnet.sigmoid(T.dot(x_t, self.w_xo) +
            ####                     T.dot(h_tm1, self.w_ho) +
            ####                     T.dot(c_t, self.w_co) +
            ####                     self.b_o)
            h_t = o_t * T.tanh(c_t)
            return [c_t, h_t]

        # If we use batches, we have to permute the first and second dimension.
        if self.with_batch:
            self.input = input.dimshuffle(1, 0, 2)
            outputs_info = [T.alloc(x, self.input.shape[1], self.hidden_dim)
                            for x in [self.c_0, self.h_0]]
        else:
            ####self.input = input
            self.input = T.dot(input, self.w_x) + self.b
            outputs_info = [self.c_0, self.h_0]

        [_, h], _ = theano.scan(
            fn=recurrence,
            sequences=self.input,
            outputs_info=outputs_info,
            n_steps=self.input.shape[0]
        )
        self.h = h
        self.output = h[-1]

        return self.output
```
